indc.go

```
@@ -202,6 +202,225 @@ func (a Aroon) namedMarshalJSON() ([]byte, error) {
+
+	return bb, nil
+}
+
+// Equal checks whether provided band has exactly the same values as main
+// band.
+func (bb BB) Equal(bb1 BB) bool {
+	if bb.valid != bb1.valid || bb.percent != bb1.percent ||
+		bb.band != bb1.band || !bb.stdDevs.Equal(bb1.stdDevs) ||
+		bb.length != bb1.length || bb.offset != bb1.offset {
+
+		return false
+	}
+
+	return true
+}
+
+func (bb BB) equal(i Indicator) bool {
+	b1, ok := i.(BB)
+	if ok {
+		return bb.Equal(b1)
+	}
+
+	return ok
+}
+
+// Percent returns percent configuration option.
+func (bb BB) Percent() bool {
+	return bb.percent
+}
+
+// Band returns band configuration option.
+func (bb BB) Band() Band {
+	return bb.band
+}
+
+// StandardDeviations returns standard deviation configuration option.
+func (bb BB) StandardDeviations() decimal.Decimal {
+	return bb.stdDevs
+}
+
+// Length returns length configuration option.
+func (bb BB) Length() int {
+	return bb.length
+}
+
+// Offset returns offset configuration option.
+func (bb BB) Offset() int {
+	return bb.offset
+}
+
+// validate checks whether BB was configured properly or not.
+func (bb *BB) validate() error {
+	if err := bb.band.Validate(); err != nil {
+		return err
+	}
+
+	if bb.percent && bb.band == BandMiddle || bb.percent && bb.band == BandWidth {
+		return errors.New("invalid bb configuration")
+	}
+
+	if bb.length < 1 {
+		return ErrInvalidLength
+	}
+
+	if bb.offset < 0 {
+		return ErrInvalidOffset
+	}
+
+	bb.valid = true
+
+	return nil
+}
+
+// Calc calculates BB from the provided data points slice.
+// Calculation is based on formula provided by investopedia.
+// https://www.investopedia.com/terms/b/bollingerbands.asp.
+// All credits are due to John Bollinger who developed BB indicator.
+func (bb BB) Calc(dd []decimal.Decimal) (decimal.Decimal, error) {
+	if !bb.valid {
+		return decimal.Zero, ErrInvalidIndicator
+	}
+
+	dd, err := resize(dd, bb.Count()-bb.offset, bb.offset)
+	if err != nil {
+		return decimal.Zero, err
+	}
+
+	s, _ := NewSMA(bb.length, 0)
```

I don't think we should discard error values, even if they seem unlikely to be returned.